R/MsExperiment.R

```diff
@@ -214,48 +214,99 @@ setMethod("readMsObject", signature(object = "MsExperiment",
-
-              assay_data <- read.table(file.path(url, selected_assay),
-                                       header = TRUE, sep = "\t")
-              assay_data$injection_index <- seq_len(nrow(assay_data))
+              }
+              assay_data <- read.table(paste0(pth, selected_assay),
+                                       header = TRUE, sep = "\t",
+                                       check.names = FALSE)
 
               ## Extract and read sample info files
-              sample_info_files <- grep("s_", files, value = TRUE)
-              sample_info <- read.table(file.path(url, sample_info_files),
-                                        header = TRUE, sep = "\t")
-              merged_data <- merge(assay_data, sample_info, by = "Sample.Name")
-              merged_data <- merged_data[order(merged_data$injection_index), ]
-              merged_data <- merged_data[, colSums(is.na(merged_data)) <
-                                             nrow(merged_data)]
+              s_files <- all_fls[grepl("^s_", all_fls)]
+              sample_info <- read.table(paste0(pth, s_files),
+                                        header = TRUE, sep = "\t",
+                                        check.names = FALSE)
+
+              # merging
+              ord <- match(assay_data$`Sample Name`, sample_info$`Sample Name`)
+              merged_data <- cbind(assay_data, sample_info[ord, ])
+              names(merged_data) <- gsub(" ", "_", names(merged_data))
+              if (keepProtocol || keepOntology || simplify)
+                  merged_data <- .clean_merged(x = merged_data,
+                                               keepProtocol = keepProtocol,
+                                               keepOntology = keepOntology,
+                                               simplify = simplify)
+
+              ## Assemble object
+              object@spectra <- Spectra::Spectra(mtblsId = param@mtblsId,
+                                                 source = MsBackendMetaboLights::MsBackendMetaboLights(),
+                                                 assayName = selected_assay,
+                                                 filePattern = param@filePattern)
+
+              ## sample to spectra link
+              fl <- object@spectra@backend@spectraData[1, "derived_spectral_data_file"]
+              nme <- colnames(merged_data)[which(merged_data[1, ] == fl)]
+              merged_data <- merged_data[grepl(param@filePattern,
+                                               merged_data[, nme]), ]
 
               object@sampleData <- DataFrame(merged_data)
+              object <- MsExperiment::linkSampleData(object,
+                                                     with = paste0("sampleData.",
+                                                                nme,
+                                                                "= spectra.derived_spectral_data_file"))
```

Comment:
@jorainer You know these functions better than me, tell me if there is a better way to do this :)